### PR TITLE
Type hints and documentation

### PR DESCRIPTION
**Pull request by MaxBlesch**
This PR is used to collect all the work from @FelixAhlbrecht and use his work to build a consistent documentation of types, input variables and outputs.

Tasks: 

- [x] Type hints. @FelixAhlbrecht 

- [x]  Reference inputs to variable and function documentation. @FelixAhlbrecht 

- [x]  Create param files documentation. @MaxBlesch 

- [x]  Create internal type documentation. @MaxBlesch 

- [x] Check on type hints consistency. @MaxBlesch 

- [x]  Check on type consitencies. @MaxBlesch

### CODE REVIEW

File: gettsim/benefits/arbeitsl_geld_2/arbeitsl_geld_2.py [hunk 1]
```diff
@@ -100,13 +145,25 @@ def regelsatz_m_hh_bis_2010(
...
+    Returns
+    -------
+
+    """
     data = np.where(
         anz_erwachsene_hh == 1,
         arbeitsl_geld_2_params["regelsatz"] * (1 + alleinerziehenden_mehrbedarf_hh),
         arbeitsl_geld_2_params["regelsatz"]
         * (
-            (2 + alleinerziehenden_mehrbedarf_hh)
```
Comment (MaxBlesch) [inline gettsim/benefits/arbeitsl_geld_2/arbeitsl_geld_2.py:108]: I think this is wrong? We do not need to add the mehrbedarf if the household contains both parents? @Eric-Sommer  Our tests don't fail if I delete that! We should have a test case detecting that.

**Comment by tobiasraabe** [comment]
Please add https://numpydoc.readthedocs.io/en/latest/format.html to the contribution guide for an explanation on how to write docstrings.

File: docs/gettsim_objects/functions.rst [hunk 2]
```diff
@@ -9,6 +11,7 @@ further below. By clicking on the ``source`` hyperlink next to the function sign
 you can see the full source code of the function.
 
 .. automodsumm:: gettsim.functions
```
[the remark this answers is not in the document]
Comment (ChristianZimpelmann) [reply docs/gettsim_objects/functions.rst:13]: Should we add a third column specifying the type of the output?

Comment (ChristianZimpelmann) [inline docs/gettsim_objects/functions.rst:13]: In the code documentation part, gettsim is in lower case, in contrast to GETTSIM in the rest of the documentation. That should be harmonized

File: docs/gettsim_objects/params.rst [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+dictionaries. The :code:`params` argument created in :func:`set_up_policy_environment`
+and then used in :func:`compute_taxes_and_transfers` is a dictionary of these
+compartment specific dictionaries. Thus it is a dictionary of dictionaries, where the
+keys are the name of the compartment. The files from which
+:func:`set_up_policy_environment` loads the default paramters for each year, can be
+found `here <https://github.com/iza-institute-of-labor-economics/gettsim/tree/master
+/gettsim/data>`_ and are named by the compartment. Our current guidlines on how these
+files are set up can be found in the GEP-2.
+
+Inside gettsim the functions don't operate with the full The :code:`params` file and
```
Comment (ChristianZimpelmann) [inline docs/gettsim_objects/params.rst:14]: "The" too much (also some lines below)

File: docs/gettsim_objects/params.rst [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+instead just use each compartment specific parameter dictionary. The names of these
+dictionaries are named by the structure :code:`compartment_name + _params`. The explicit
+names can be found in the table below. gettsim reads the function arguments and
+selects the corresponding part of the The :code:`params` file. By construction the
+:code:`compartment_name + _params` have to be the last arguments.
```
Comment (ChristianZimpelmann) [inline docs/gettsim_objects/params.rst:19]: have to be the last argument when doing what/when calling which function?

File: gettsim/interface.py [hunk 5]
```diff
@@ -117,6 +122,43 @@ def compute_taxes_and_transfers(
...
+        A dictionary of functions.
+
+    Returns
+    -------
+    ValueError
+        Fail if the data types are not matching the required in gettsim.
+
+    """
+    check_data = True
+    for column_name, series in data.items():
```
Comment (ChristianZimpelmann) [inline gettsim/interface.py]: wusste gar nicht, dass `.items()` auch mit einem DataFrame funktioniert.. cool

File: gettsim/interface.py [hunk 6]
```diff
@@ -117,6 +122,43 @@ def compute_taxes_and_transfers(
...
+    functions : dict of callable
+        A dictionary of functions.
+
+    Returns
+    -------
+    ValueError
+        Fail if the data types are not matching the required in gettsim.
+
+    """
+    check_data = True
```
Comment (ChristianZimpelmann) [inline gettsim/interface.py]: That needs to go in the for-loop if I am not mistaken

File: gettsim/interface.py [hunk 7]
```diff
@@ -117,6 +122,43 @@ def compute_taxes_and_transfers(
...
+    for column_name, series in data.items():
+        if column_name in STANDARD_DATA_TYPES:
+            internal_type = STANDARD_DATA_TYPES[column_name]
+            check_data = check_if_series_has_internal_type(series, internal_type)
+        elif column_name in columns_overriding_functions:
+            internal_type = functions[column_name].__annotations__["return"]
```
Comment (ChristianZimpelmann) [inline gettsim/interface.py]: We should take care of elements in `columns_overriding_functions` that don't correspond to a function: either ignore those or throw an explicit error (I suggest the latter)

File: gettsim/tests/test_docs.py [hunk 8]
```diff
@@ -21,3 +26,51 @@ def test_funcs_in_doc_module_and_func_from_internal_files_are_the_same():
     }
 
     assert set(documented_functions) == set(internal_functions)
+
+
+def test_type_hints():
```
Comment (ChristianZimpelmann) [inline gettsim/tests/test_docs.py]: add a short docstring

File: gettsim/tests/test_tax_transfer.py [hunk 9]
```diff
@@ -77,3 +40,44 @@ def test_tax_transfer(
         functions=policy_functions,
         targets=DEFAULT_TARGETS,
     )
+
+
+@pytest.mark.parametrize("year", YEARS)
```
Comment (ChristianZimpelmann) [inline gettsim/tests/test_tax_transfer.py]: Looks good.

But should we also include a test for `_fail_if_datatype_is_false`?

File: gettsim/typing.py [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+    out : bool
+        Return check variable.
+    """
+    dtype = series.dtype
+    if internal_type == FloatSeries:
+        # I think we can allow for ints here?!
+        out = np.issubdtype(dtype, np.integer) or np.issubdtype(dtype, np.floating)
+    elif internal_type == BoolSeries:
+        out = np.issubdtype(dtype, np.bool)
+    elif internal_type == IntSeries:
```
Comment (ChristianZimpelmann) [inline gettsim/typing.py]: Make sure this covers both the old integer type and the new (experimental) pandas integer type (https://pandas.pydata.org/pandas-docs/stable/user_guide/integer_na.html). Not sure if it does right now.

Same question for the BoolSeries

**Review by ChristianZimpelmann — CHANGES_REQUESTED** [review]
Great work! Thank you very much!

The documentation looks much better both contentwise and design-/structurewise.

I have a bunch of comments/requests though. Feel free to disagree!

File: docs/how_to_guides/visualizing_the_system.ipynb [hunk 11]
```diff
@@ -953,384 +876,22 @@
...
-      "application/vnd.bokehjs_exec.v0+json": ""
-     },
-     "metadata": {
-      "application/vnd.bokehjs_exec.v0+json": {
-       "id": "1451"
-      }
-     },
-     "output_type": "display_data"
+     "ename": "ValueError",
+     "evalue": "The following selected nodes are not in the DAG:\n\n[\n    \"zu_verst_eink_kinderfreib_tu\",\n]\n",
```
[the remark this answers is not in the document]
Comment (MaxBlesch) [reply docs/how_to_guides/visualizing_the_system.ipynb:880]: Good catch. I didn't ran the how-to guides. Error is gone!

File: gettsim/interface.py (hunk 6 above)
Comment (MaxBlesch) [reply gettsim/interface.py:144]: I think it doesn't matter. If once check_data is set to False it stops and raises the error....Oh well maybe for the debug mode. Changed it!

File: gettsim/tests/test_docs.py (hunk 8 above)
Comment (MaxBlesch) [reply gettsim/tests/test_docs.py:31]: Done

File: gettsim/typing.py (hunk 10 above)
Comment (MaxBlesch) [reply gettsim/typing.py:33]: I changed to pandas checks throughout!

File: gettsim/tests/test_tax_transfer.py (hunk 9 above)
Comment (MaxBlesch) [reply gettsim/tests/test_tax_transfer.py:45]: Also wir haben das bis jetzt für keine fal function und insgesamt eigl kaum uni tests. Könnte das natürlich machen, aber es ist ja auch am Anfang oft gefailed, weil ich es benutzt habe um die type hints coherent zu machen. Im Prinzip wird es ja auch durch unsere test daten gecheckt. Soll ich trotzdem?

File: gettsim/interface.py (hunk 7 above)
Comment (MaxBlesch) [reply gettsim/interface.py:150]: There is already `_fail_if_columns_overriding_functions_are_not_in_functions`. It was just not called. Will change!

File: docs/index.rst [hunk 12]
```diff
@@ -19,14 +19,8 @@ Welcome to GETTSIM's documentation!
    installation
```
[the remark this answers is not in the document]
Comment (MaxBlesch) [reply docs/index.rst:19]: I just changed the range of lines we import. Now the installation information can solely be found in installation.

File: docs/tutorials/policy_functions.ipynb [hunk 13]
```diff
@@ -15,35 +15,7 @@
    "cell_type": "code",
    "execution_count": 1,
    "metadata": {},
-   "outputs": [
-    {
```
[the remark this answers is not in the document]
Comment (MaxBlesch) [reply docs/tutorials/policy_functions.ipynb:19]: In my eyes, this what's in there. But maybe @amageh has there a better overview!

File: docs/gettsim_objects/functions.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (MaxBlesch) [reply docs/gettsim_objects/functions.rst:13]: That is not that easy. Maybe I create an issue formatting function documentation table?

File: docs/index.rst [hunk 14]
```diff
@@ -19,14 +19,8 @@ Welcome to GETTSIM's documentation!
...
    tutorials/index
    how_to_guides/index
-   crosswalk
```
[the remark this answers is not in the document]
Comment (MaxBlesch) [reply docs/index.rst:22]: Create issue?

File: docs/gettsim_objects/functions.rst [hunk 15]
```diff
@@ -9,6 +11,7 @@ further below. By clicking on the ``source`` hyperlink next to the function sign
...
+   :functions-only:
```
[the remark this answers is not in the document]
Comment (MaxBlesch) [reply docs/gettsim_objects/functions.rst:14]: Create issue: Polish function table?

File: docs/index.rst (hunk 14 above)
[the remark this answers is not in the document]
Comment (MaxBlesch) [reply docs/index.rst:22]: I think your first comment is adressed or at least I read this sentence in the documentation. Or do you mean to write more? For the second I would open an issue.

File: docs/tutorials/policy_functions.ipynb (hunk 13 above)
[the remark this answers is not in the document]
Comment (amageh) [reply docs/tutorials/policy_functions.ipynb:19]: True, seems to have slipped by us. I'll just add a sentence linking to the policy functions how-to guide somewhere at the beginning of the tutorial. Can I do that on this PR @MaxBlesch?

File: docs/index.rst (hunk 14 above)
[the remark this answers is not in the document]
Comment (ChristianZimpelmann) [reply docs/index.rst:22]: I meant having one explanatory sentence on the starting page already (there, I don't see it)

[the remark this answers is not in the document]
Comment (ChristianZimpelmann) [reply docs/index.rst:22]: Will create issue for respy design

[the remark this answers is not in the document]
Comment (MaxBlesch) [reply docs/index.rst:22]: > I meant having one explanatory sentence on the starting page already (there, I don't see it)
Ah okay. Will do!

File: docs/gettsim_objects/functions.rst (hunk 15 above)
[the remark this answers is not in the document]
Comment (ChristianZimpelmann) [reply docs/gettsim_objects/functions.rst:14]: is created

File: docs/tutorials/policy_functions.ipynb (hunk 13 above)
[the remark this answers is not in the document]
Comment (MaxBlesch) [reply docs/tutorials/policy_functions.ipynb:19]: Thanks @amageh for the help. I included a sentence after the first two paragraphs in the notebook!

**Review by ChristianZimpelmann — APPROVED** [review]

File: gettsim/benefits/arbeitsl_geld_2/arbeitsl_geld_2.py (hunk 1 above)
Comment (hendshe) [reply gettsim/benefits/arbeitsl_geld_2/arbeitsl_geld_2.py:108]: Just to confirm, the model can identify the "Mehrbedarf" of single parents only . None of the other additional needs cases could be identified in it (e.g. employable disabled people taking part in labor market reintegration measures)? I understand that all other cases could not be easily identified, just wanted to make sure it is the case. I have only seen it for single parents in the policy functions.

**Pull request merged**